File: app/src/main/java/com/farzin/testdnschanger/service/ConnectivityBackgroundService.java
package com.farzin.testdnschanger.service;

import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.VpnService;
import android.os.Build;
import android.os.IBinder;

import androidx.annotation.Nullable;


public class ConnectivityBackgroundService extends Service {
    private BroadcastReceiver connectivityChange = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            boolean connected = !intent.hasExtra("noConnectivity");
            int type = intent.getIntExtra("networkType", -1);
            if(!connected)return;
            if(type == ConnectivityManager.TYPE_WIFI){
                startService();
            }else if(type == ConnectivityManager.TYPE_MOBILE ){
                startService();
            }
        }
    };

    private void startService(){
        Intent i = VpnService.prepare(this);
        if (i == null) {
            Intent serviceIntent = new Intent(this, DNSVpnService.class);
            serviceIntent.putExtra("start_vpn", true);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                startForegroundService(serviceIntent);
            } else {
                startService(serviceIntent);
            }
        } else {
            BackgroundVpnConfigureActivity.startBackgroundConfigure(this, true);
        }
    }


    @Override
    public void onCreate() {
        super.onCreate();
        registerReceiver(connectivityChange, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
        ConnectivityManager cm =(ConnectivityManager)getSystemService(Context.CONNECTIVITY_SERVICE);

        NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
        if(activeNetwork == null)return;
        if(activeNetwork.getType() == ConnectivityManager.TYPE_WIFI ){
            startService();
        }else if(activeNetwork.getType() == ConnectivityManager.TYPE_MOBILE ){
            startService();
        }
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: app/src/main/java/com/farzin/testdnschanger/service/ConnectivityBackgroundService.kt
package com.farzin.testdnschanger.service

import android.app.Service
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.net.VpnService
import android.os.Build
import android.os.IBinder

class ConnectivityBackgroundService : Service() {
    private lateinit var connectivityManager: ConnectivityManager
    private lateinit var networkCallback: ConnectivityManager.NetworkCallback

    override fun onCreate() {
        super.onCreate()
        connectivityManager = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        networkCallback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                val capabilities = connectivityManager.getNetworkCapabilities(network) ?: return
                if (isWifiOrMobile(capabilities)) {
                    startService()
                }
            }

            override fun onLost(network: Network) {
                // Handle network loss if needed
            }
        }
        val networkRequest = NetworkRequest.Builder()
            .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
            .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
            .build()
        connectivityManager.registerNetworkCallback(networkRequest, networkCallback)
    }

    private fun isWifiOrMobile(capabilities: NetworkCapabilities): Boolean {
        return capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
                capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)
    }

    private fun startService() {
        val i = VpnService.prepare(this)
        if (i == null) {
            val serviceIntent = Intent(this, DNSVpnService::class.java)
            serviceIntent.putExtra("start_vpn", true)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                startForegroundService(serviceIntent)
            } else {
                startService(serviceIntent)
            }
        } else {
            // deleted class
//            startBackgroundConfigure(this, true)
        }
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        return START_STICKY
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onDestroy() {
        super.onDestroy()
        connectivityManager.unregisterNetworkCallback(networkCallback)
    }
}

